### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by rodrieiz**
https://trello.com/c/dWAQXtW5/2-5-as-a-user-i-should-be-able-to-sign-up-so-my-account-is-created-in-target-mvd

Sign up using devise_token_auth

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (enzofab91) [inline spec/models/user_spec.rb]: What about adding a test to the model validations?

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class User < ActiveRecord::Base
+  extend Devise::Models
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+  include DeviseTokenAuth::Concerns::User
+  # Validate params
+  validates :gender, presence: true, inclusion: { in: ["Male","Female"]}
```
Comment (enzofab91) [inline app/models/user.rb]: Maybe validate uniqueness of email. WDYT?

See https://guides.rubyonrails.org/active_record_validations.html#uniqueness

**Review by enzofab91 — COMMENTED** [review]
Nice work! 👏 

Left some comments. It would be nice to have a test to validate the sign up endpoint.

File: app/models/user.rb (hunk 2 above)
Comment (rodrieiz) [reply app/models/user.rb]: Yes, of course!

File: spec/models/user_spec.rb (hunk 1 above)
Comment (rodrieiz) [reply spec/models/user_spec.rb]: I´ll try to do it!

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,2 +1,3 @@
 class ApplicationController < ActionController::API
+        include DeviseTokenAuth::Concerns::SetUserByToken
```
Comment (sebastiancaraballo) [inline app/controllers/application_controller.rb]: Be careful with this indentation.

**Review by sebastiancaraballo — COMMENTED** [review]
Please add some tests for the sign-up endpoint

**Review by enzofab91 — APPROVED** [review]
Good work!

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class User < ActiveRecord::Base
+  extend Devise::Models
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+  include DeviseTokenAuth::Concerns::User
+  # Validate params
```
Comment (sebastiancaraballo) [inline app/models/user.rb]: Avoid comments

File: spec/factories/users.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :user do
+    email {Faker::Internet.email}
+    gender {"Female"}
+    password {"123456"}
+  end
+end
```
Comment (sebastiancaraballo) [inline spec/factories/users.rb]: Blank line

File: spec/models/user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  
+  before(:all) do 
+    @user1 = create(:user)
+  end
```
Comment (sebastiancaraballo) [inline spec/models/user_spec.rb]:
```suggestion
  let!(:user) { create(:user) }
```

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    expect(@user1).to be_valid
+  end
+
+  it "invalid user attributes, duplicate email" do
+    @user2 = build(:user , email: @user1.email)
+    expect(@user2).to_not be_valid
+  end
+
+  it "invalid user attributes, incorrect gender" do
+    @user3 = build(:user, gender: "GenderTest")
```
Comment (sebastiancaraballo) [inline spec/models/user_spec.rb]: Try using `let` blocks.

File: spec/requests/registration_spec.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      {
+        email: 'test1@test.com.uy',
+        gender: 'Female',
+        password: 'abcd1234',
+        password_confirmation: 'abcd1234'
+      }
+    }
+  end
+
+  context 'with correct params' do
```
Comment (sebastiancaraballo) [inline spec/requests/registration_spec.rb:5]: Maybe add a context for invalid params.

**Review by sebastiancaraballo — COMMENTED** [review]
Nice. Please add a Github workflow to check tests and code analysis tools

File: config/routes.rb [hunk 9]
```diff
@@ -1,3 +1,7 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  mount_devise_token_auth_for 'User', at: 'registration', controllers: {registrations: "registrations"}
+
+  scope :api, defaults: { format: :json} do
```
Comment (DJA89) [inline config/routes.rb]: Be careful with the spaces inside `{}`. We recommend using https://github.com/rubocop/rubocop which will tell you this things, there are also plug ins for some editors so you don't even have to run rubocop by hand

File: spec/factories/users.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :user do
+    email {Faker::Internet.email}
+    gender {"Female"}
+    password {"123456"}
```
Comment (DJA89) [inline spec/factories/users.rb]: Add a space inside `{}` at the beggining and at the end, like so:
`email {Faker::Internet.email}` -> `email { Faker::Internet.email }`

File: config/routes.rb (hunk 9 above)
Comment (rodrieiz) [reply config/routes.rb]: Thanks, I´ve just added the plugin for vscode

File: spec/models/user_spec.rb (hunk 7 above)
Comment (rodrieiz) [reply spec/models/user_spec.rb]: This is your suggestion?
let(:user3) {build(:user), gender: 'GenderTest'}

Comment (santir489) [reply spec/models/user_spec.rb]: And maybe you can test models validation like the following:
https://github.com/thoughtbot/shoulda-matchers#should-vs-is_expectedto

**Review by sebastiancaraballo — APPROVED** [review]

**Pull request merged**